packages/react-meteor-data/useTracker.ts

```diff
@@ -178,107 +95,72 @@ const useTrackerNoDeps: useTrackerSignature = (reactiveFn, deps = null, computat
     forceUpdate();
 
     // stop the computation on unmount
-    return () => dispose(refs);
+    return () =>{
+      refs.computation?.stop();
+    }
   }, []);
 
   return refs.trackerData;
 }
 
-const useTrackerWithDeps: useTrackerSignature = (reactiveFn, deps: Array<any>, computationHandler) => {
-  const { current: refs } = useRef<TrackerRefs>({
-    reactiveFn,
-    isMounted: false,
-    trackerData: null
-  });
-  const forceUpdate = useForceUpdate();
-
-  // Always have up to date deps and computations in all contexts
-  refs.reactiveFn = reactiveFn;
-  refs.deps = deps;
-  if (computationHandler) {
-    refs.computationHandler = computationHandler;
-  }
+const useTrackerWithDeps = <T = any>(reactiveFn: IReactiveFn<T>, deps: DependencyList): T => {
+  let [data, setData] = useState<T>();
 
-  // We are abusing useMemo a little bit, using it for it's deps
-  // compare, but not for it's memoization.
   useMemo(() => {
-    // stop the old one.
-    dispose(refs);
-
-    track(refs, forceUpdate, tracked)
-
-    // Tracker creates side effect in render, which can be problematic in some cases, such as
-    // Suspense or concurrent rendering or if an error is thrown and handled by an error boundary.
-    // We still want synchronous rendering for a number of reasons (see readme). useTracker works
-    // around memory/resource leaks by setting a time out to automatically clean everything up,
-    // and watching a set of references to make sure everything is choreographed correctly.
-    if (!refs.isMounted) {
-      // Components yield to allow the DOM to update and the browser to paint before useEffect
-      // is run. In concurrent mode this can take quite a long time. 1000ms should be enough
-      // in most cases.
-      refs.disposeId = setTimeout(() => {
-        if (!refs.isMounted) {
-          dispose(refs);
-        }
-      }, 1000);
+    // To jive with the lifecycle interplay between Tracker/Subscribe, run the
+    // reactive function in a computation, then stop it, to force flush cycle.
+    const comp = Tracker.nonreactive(
+      () => Tracker.autorun((c: Tracker.Computation) => {
+        if (c.firstRun) data = reactiveFn();
```

Comment:
Why are we not using `setData(reactiveFn())` here as it is done below?

Reply:
Because that would cause an immediate rerender. The idea here is to temporarily override the value of data during the initial render (or immediately after deps change). Then when the component is mounted the next tracker instance set up in `useEffect` will update the state data.

There is an assumption made here, which I'm not confident in, that the initial render will never be run more than once before useEffect is run. I'm thinking that might not be true if a user invokes an immediate rerender themselves (by triggering some hook in render). I can actually write a test for this.

I already have a mitigation in place for this possibility, but I don't want to push it until I've confirmed it's needed.
https://github.com/meteor/react-packages/blob/better-handle-data-with-refs/packages/react-meteor-data/useTracker.ts

Reply:
found an issue https://github.com/meteor/react-packages/issues/327

Reply:
One of the things I changed in #321 was to keep the value of the first render in a ref - can you try that branch and see if it solves your issue?